refactor: Add `Require` node to generic AST

nmote opened the pull request:
Currently, in JS, `require('foo')` is represented using `OtherExpr` for the `require` part of that expression. This is fine currently since we don't do any analysis of `require` calls after the AST generic translation. Instead, we desugar them to `ImportFrom` in some cases, and Semgrep's existing naming logic for imported symbols takes it from there.

However, that desugaring is causing some issues. See #6532 for details.

Instead of desugaring, we need to represent `Require` in the AST properly, and analyze it directly. We could simply treat it like a normal function call to a function named `require`, but it's a separate node in the JS AST and doing analysis by matching on specific literal values is more error-prone.

Test plan: Automated tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

aryx commented on semgrep-core/src/core/ast/AST_generic.ml:
```diff
@@ -748,6 +748,7 @@ and special =
   | Op of operator
   (* less: should be lift up and transformed in Assign at stmt level *)
   | IncrDecr of (incr_decr * prefix_postfix)
+  | Require
```
Can you add a comment explaining why we need this construct given we already have ImportFrom?
Maybe something like "This is similar to ImportFrom, but at the expression level; in languages such as Javascript or Ruby import/require can appear at any places, not just at the toplevel".

nmote committed:
refactor: Add `Require` node to generic AST

Currently, in JS, `require('foo')` is represented using `OtherExpr` for
the `require` part of that expression. This is fine currently since we
don't do any analysis of `require` calls after the AST generic
translation. Instead, we desugar them to `ImportFrom` in some cases, and
Semgrep's existing naming logic for imported symbols takes it from
there.

However, that desugaring is causing some issues. See #6532 for details.

Instead of desugaring, we need to represent `Require` in the AST
properly, and analyze it directly. We could simply treat it like a
normal function call to a function named `require`, but it's a separate
node in the JS AST and doing analysis by matching on specific literal
values is more error-prone.

Test plan: Automated tests
